lib/puppet/provider/package/snap.rb

```diff
@@ -7,109 +7,100 @@
-
-    raise Puppet::Error, "Could not find installed snaps (code: #{res['status-code']})" unless [200, 404].include?(res['status-code'])
+  def modify_snap(action, options = nil)
+    req = self.class.generate_request(action, determine_channel(options), options)
+    response = PuppetX::Snap::API.post("/v2/snaps/#{@resource[:name]}", req)
+    change_id = PuppetX::Snap::API.get_id_from_async_req(response)
+    PuppetX::Snap::API.complete(change_id)
+  end
 
-    res['result'].map { |hash| hash.slice('name', 'version') } if res['status-code'] == 200
+  def determine_channel(options = nil)
+    channel = self.class.channel_from_ensure(@resource[:ensure])
+    channel ||= self.class.channel_from_options(options) if options
+    channel ||= 'latest/stable'
+    channel
   end
 
-  def self.generate_request(action, options)
+  def self.generate_request(action, channel, options)
     request = { 'action' => action }
+    request['channel'] = channel unless channel.nil?
 
     if options
-      channel = parse_channel(options)
-      request['channel'] = channel unless channel.nil?
-
-      # classic, devmode and jailmode params are only available for install, refresh, revert actions.
-      if %w[install refresh revert].include?(action)
+      # classic, devmode and jailmode params are only
+      # available for install, refresh, revert actions.
+      case action
+      when 'install', 'refresh', 'revert'
         request['classic'] = true if options.include?('classic')
         request['devmode'] = true if options.include?('devmode')
         request['jailmode'] = true if options.include?('jailmode')
+      when 'remove'
+        request['purge'] = true if options.include?('purge')
       end
-
-      request['purge'] = true if action == 'remove' && options.include?('purge')
     end
 
     request
   end
 
-  def self.modify_snap(action, name, options = nil)
-    req = generate_request(action, options)
-    response = PuppetX::Snap::API.post("/v2/snaps/#{name}", req)
-    change_id = PuppetX::Snap::API.get_id_from_async_req(response)
-    PuppetX::Snap::API.complete(change_id)
+  def self.channel_from_ensure(value)
+    value = value.to_s
+    case value
+    when 'present', 'absent', 'purged', 'installed', 'latest'
+      nil
+    else
+      value
+    end
   end
 
-  def self.parse_channel(options)
-    if (channel = options.find { |e| %r{channel} =~ e })
-      return channel.split('=')[1]
+  def self.channel_from_options(options)
+    options&.find { |e| %r{channel} =~ e }&.split('=')&.last&.tap do |ch|
+      Puppet.warning("Install option 'channel' is deprecated, use ensure => '#{ch}' instead.")
     end
+  end
+
+  def self.installed_snaps
+    res = PuppetX::Snap::API.get('/v2/snaps')
+    raise Puppet::Error, "Could not find installed snaps (code: #{res['status-code']})" unless [200, 404].include?(res['status-code'])
 
-    nil
+    res['status-code'] == 200 ? res['result'].map { |hash| hash.slice('name', 'version') } : []
```

Comment:
You need to return the channel in order to be available for the `self.instances` method
```suggestion
    res['status-code'] == 200 ? res['result'].map { |hash| hash.slice('name', 'tracking-channel') } : []
```